### PR TITLE
soc: arm: stm32wb power: implement HSEM for power management

### PR DESCRIPTION
**Pull request by ABOSTM**
soc: arm: stm32wb power: implement HSEM for power management

Implementation of hardware semaphore algorithm of STM32 AN5289
to enter and exit low power.
Similar to implementation proposed in STM32WB Cube Application BLE_HeartRate:
https://github.com/STMicroelectronics/STM32CubeWB/tree/master/Projects/P-NUCLEO-WB55.Nucleo/Applications/BLE/BLE_HeartRate

### CODE REVIEW

File: soc/arm/st_stm32/stm32wb/power.c [hunk 1]
```diff
@@ -14,10 +14,25 @@
...
 
 #include <logging/log.h>
 LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
 
+/**
+  * @brief Switch the system clock on HSI
+  * @param none
+  * @retval none
+  */
+static void Switch_On_HSI(void)
```
Comment (erwango) [inline soc/arm/st_stm32/stm32wb/power.c:27]:
```suggestion
static void switch_on_hsi(void)
```

File: soc/arm/st_stm32/stm32wb/power.c [hunk 2]
```diff
@@ -14,10 +14,25 @@
...
+{
+	LL_RCC_HSI_Enable();
+	while (!LL_RCC_HSI_IsReady());
+	LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_HSI);
+	LL_RCC_SetSMPSClockSource(LL_RCC_SMPS_CLKSOURCE_HSI);
+	while (LL_RCC_GetSysClkSource() != LL_RCC_SYS_CLKSOURCE_STATUS_HSI);
```
Comment (erwango) [inline soc/arm/st_stm32/stm32wb/power.c:33]:
```suggestion
	while (!LL_RCC_HSI_IsReady()) {
	}
	
	LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_HSI);
	LL_RCC_SetSMPSClockSource(LL_RCC_SMPS_CLKSOURCE_HSI);
	while (LL_RCC_GetSysClkSource() != LL_RCC_SYS_CLKSOURCE_STATUS_HSI) {
	}
```

File: soc/arm/st_stm32/stm32wb/power.c [hunk 3]
```diff
@@ -26,12 +41,31 @@ void pm_power_state_set(struct pm_state_info info)
 		return;
 	}
 
+	/* Implementation of STM32 AN5289, proposed in STM32WB Cube Application
+	 * BLE_HeartRate
+	 * https://github.com/STMicroelectronics/STM32CubeWB/tree/master/Projects/P-NUCLEO-WB55.Nucleo/Applications/BLE/BLE_HeartRate
```
Comment (erwango) [inline soc/arm/st_stm32/stm32wb/power.c:46]: If that's implementation of STM32 AN5289, is there a need to give Cube sample reference ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: soc/arm/st_stm32/stm32wb/power.c (hunk 1 above)
Comment (ABOSTM) [reply soc/arm/st_stm32/stm32wb/power.c]: Done

File: soc/arm/st_stm32/stm32wb/power.c (hunk 2 above)
Comment (ABOSTM) [reply soc/arm/st_stm32/stm32wb/power.c]: Done

File: soc/arm/st_stm32/stm32wb/power.c (hunk 3 above)
Comment (ABOSTM) [reply soc/arm/st_stm32/stm32wb/power.c:49]: I prefer to keep both reference: AN5289 is the theoretical one, Cube example is the practical one (from where I copy/past :smiley:  )

Comment (ABOSTM) [reply soc/arm/st_stm32/stm32wb/power.c]: Cube example reference removed from code.
And added to PR description.

**Comment by ABOSTM** [comment]
Buildkite issue is known:
https://github.com/zephyrproject-rtos/zephyr/issues/36713

**Review by erwango — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]

**Pull request merged**